internal/controller/rotators/aws_oidc_rotator.go

```diff
@@ -0,0 +1,199 @@
+	backendSecurityPolicyNamespace string,
+	backendSecurityPolicyName string,
+	preRotationWindow time.Duration,
+	region string,
+) (*AWSOIDCRotator, error) {
+	cfg, err := defaultAWSConfig(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load AWS config: %w", err)
+	}
+
+	cfg.Region = region
+
+	if proxyURL := os.Getenv("AI_GATEWAY_STS_PROXY_URL"); proxyURL != "" {
+		cfg.HTTPClient = &http.Client{
+			Transport: &http.Transport{
+				Proxy: func(*http.Request) (*url.URL, error) {
+					return url.Parse(proxyURL)
+				},
+			},
+		}
+	}
+
+	stsClient := NewSTSClient(cfg)
+
+	return &AWSOIDCRotator{
+		client:                         client,
+		kube:                           kube,
+		logger:                         logger.WithName("aws-oidc-rotator"),
+		stsOps:                         stsClient,
+		backendSecurityPolicyNamespace: backendSecurityPolicyNamespace,
+		backendSecurityPolicyName:      backendSecurityPolicyName,
+		preRotationWindow:              preRotationWindow,
+	}, nil
+}
+
+// SetSTSOperations sets the STS operations implementation - primarily used for testing.
+func (r *AWSOIDCRotator) SetSTSOperations(ops STSClient) {
+	r.stsOps = ops
+}
+
+// IsExpired checks if the preRotation time is before the current time.
+func (r *AWSOIDCRotator) IsExpired() bool {
+	preRotationExpirationTime := r.GetPreRotationTime()
+	return IsBufferedTimeExpired(0, preRotationExpirationTime)
+}
+
+// UpdateCtx is used to update the context used in AWSOIDCRotator functions.
+// This can be used to set timeouts for outgoing calls to assume role.
+func (r *AWSOIDCRotator) UpdateCtx(ctx context.Context) {
+	r.ctx = ctx
+}
+
+// GetPreRotationTime gets the expiration time minus the preRotation interval or return zero value for time.
+func (r *AWSOIDCRotator) GetPreRotationTime() time.Time {
+	secret, err := LookupSecret(r.ctx, r.client, r.backendSecurityPolicyNamespace, GetBSPSecretName(r.backendSecurityPolicyName))
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return time.Time{}
+		}
+		return time.Time{}
+	}
+	expirationTime, err := GetExpirationSecretAnnotation(secret)
+	if err != nil {
+		return time.Time{}
+	}
+	preRotationTime := expirationTime.Add(-r.preRotationWindow)
+	return preRotationTime
+}
+
+// Rotate implements the retrieval and storage of AWS sts credentials.
+func (r *AWSOIDCRotator) Rotate(region, roleARN, token string) error {
+	r.logger.Info("rotating AWS sts temporary credentials",
+		"namespace", r.backendSecurityPolicyNamespace,
+		"name", r.backendSecurityPolicyName)
+
+	var result *sts.AssumeRoleWithWebIdentityOutput
+	var err error
+
+	// This adds timeout via ctx from the caller.
+	for result == nil {
+		timer := time.NewTimer(time.Second)
+		select {
+		case <-r.ctx.Done():
+			return r.ctx.Err()
+		case <-timer.C:
+			result, err = r.assumeRoleWithToken(roleARN, token)
+			if err != nil {
+				r.logger.Error(err, "failed to assume role", "role", roleARN, "ID", token)
+				return err
+			}
```

Comment:
same comment on no idea why it's sleeping one second like this. Make `assumeRoleWithToken` accept the context and pass a context there.